fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/RecursiveCursor.java

```diff
@@ -0,0 +1,458 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.cursors;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.ByteArrayContinuation;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorContinuation;
+import com.apple.foundationdb.record.RecordCursorProto;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorStartContinuation;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+import com.apple.foundationdb.tuple.ByteArrayUtil2;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.ZeroCopyByteString;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+
+/**
+ * A cursor that flattens a tree of cursors.
+ * A root cursor seeds the output and then each output element is additionally mapped into a child cursor.
+ * @param <T> the type of elements of the cursors
+ */
+@API(API.Status.EXPERIMENTAL)
+@SuppressWarnings("PMD.CloseResource")
+public class RecursiveCursor<T> implements RecordCursor<RecursiveCursor.RecursiveValue<T>> {
+
+    @Nonnull
+    private final ChildCursorFunction<T> childCursorFunction;
+    @Nullable
+    private final Function<T, byte[]> checkValueFunction;
+    @Nonnull
+    private final List<RecursiveNode<T>> nodes;
+
+    private int currentDepth;
+
+    @Nullable
+    private RecordCursorResult<RecursiveValue<T>> lastResult;
+
+    private RecursiveCursor(@Nonnull ChildCursorFunction<T> childCursorFunction,
+                            @Nullable Function<T, byte[]> checkValueFunction,
+                            @Nonnull List<RecursiveNode<T>> nodes) {
+        this.childCursorFunction = childCursorFunction;
+        this.checkValueFunction = checkValueFunction;
+        this.nodes = nodes;
+    }
+
+    /**
+     * Create a recursive cursor.
+     * @param rootCursorFunction a function that given a continuation or {@code null} returns the children of the root
+     * @param childCursorFunction a function that given a value and a continuation returns the children of that level
```

Comment:
This is more like `START WITH ... CONNECT BY` than a recursive CTE, but it should be possible for planning to recognize common cases of the latter adaptable of the former.